airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/up_to_date/pipeline.py

```diff
@@ -0,0 +1,326 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+from __future__ import annotations
+
+import re
+from typing import TYPE_CHECKING
+
+import dagger
+import git
+import requests
+import toml
+from connector_ops.utils import ConnectorLanguage  # type: ignore
+from jinja2 import Environment, PackageLoader, select_autoescape
+from pipelines.airbyte_ci.connectors.build_image.steps.python_connectors import BuildConnectorImages
+from pipelines.airbyte_ci.connectors.bump_version.pipeline import AddChangelogEntry, BumpDockerImageTagInMetadata, get_bumped_version
+from pipelines.airbyte_ci.connectors.consts import CONNECTOR_TEST_STEP_ID
+from pipelines.airbyte_ci.connectors.context import ConnectorContext, PipelineContext
+from pipelines.airbyte_ci.connectors.reports import ConnectorReport, Report
+from pipelines.consts import LOCAL_BUILD_PLATFORM
+from pipelines.dagger.actions.python.common import with_python_connector_installed
+from pipelines.helpers.execution.run_steps import STEP_TREE, StepToRun, run_steps
+from pipelines.helpers.utils import sh_dash_c
+from pipelines.models.steps import Step, StepResult, StepStatus
+
+if TYPE_CHECKING:
+    from typing import Iterable, List, Optional
+
+    from anyio import Semaphore
+
+PACKAGE_NAME_PATTERN = r"^([a-zA-Z0-9_.\-]+)(?:\[(.*?)\])?([=~><!]=?[a-zA-Z0-9\.]+)?$"
+
+
+class CheckIsUpdateCdkCandidate(Step):
+    """Check if the connector is a candidate for migration to poetry.
+    Candidate conditions:
+    - The connector is a Python connector.
+    - The connector is a source connector.
+    - The connector is using poetry.
+    - The connector has a base image defined in the metadata.
+    """
+
+    context: ConnectorContext
+
+    title = "Check if the connector is a candidate for CDK upgrade."
+
+    def __init__(self, context: PipelineContext, cdk_version: str) -> None:
+        super().__init__(context)
+        self.cdk_version = cdk_version
+
+    async def _run(self) -> StepResult:
+        connector_dir_entries = await (await self.context.get_connector_dir()).entries()
+        if self.context.connector.language not in [ConnectorLanguage.PYTHON, ConnectorLanguage.LOW_CODE]:
+            return StepResult(
+                step=self,
+                status=StepStatus.SKIPPED,
+                stderr="The connector is not a Python connector.",
+            )
+        if self.context.connector.connector_type != "source":
+            return StepResult(
+                step=self,
+                status=StepStatus.SKIPPED,
+                stderr="The connector is not a source connector.",
+            )
+        if not "poetry.lock" in connector_dir_entries or not "pyproject.toml" in connector_dir_entries:
```

Comment:
Since the poetry.lock file is auto-generated, it feels like for this check we can assert that the connector has a pyproject.toml.

Can we add a check here that the pyproject.toml includes the CDK?

Reply:
I'll leave that TODO for my next round. I'd a bit worried about the `Candidate` notion if this pipeline does lots of things anyway. Not sure it will be helpful if there are N possible things `up_to_date` could do.